File: packages/coding-agent/src/modes/interactive/components/diff.ts
import { type Component, truncateToWidth, visibleWidth } from "@earendil-works/pi-tui";
import * as Diff from "diff";
import { getLanguageFromPath, highlightCode, theme } from "../theme/theme.ts";

/**
 * Parse diff line to extract prefix, line number, and content.
 * Format: "+123 content" or "-123 content" or " 123 content" or "     ..."
 */
function parseDiffLine(line: string): { prefix: string; lineNum: string; content: string } | null {
	const match = line.match(/^([+-\s])(\s*\d*)\s(.*)$/);
	if (!match) return null;
	return { prefix: match[1], lineNum: match[2], content: match[3] };
}

/**
 * Replace tabs with spaces for consistent rendering.
 */
function replaceTabs(text: string): string {
	return text.replace(/\t/g, "   ");
}

/**
 * Compute word-level diff and render with inverse on changed parts.
 * Uses diffWords which groups whitespace with adjacent words for cleaner highlighting.
 * Strips leading whitespace from inverse to avoid highlighting indentation.
 */
function renderIntraLineDiff(oldContent: string, newContent: string): { removedLine: string; addedLine: string } {
	const wordDiff = Diff.diffWords(oldContent, newContent);

	let removedLine = "";
	let addedLine = "";
	let isFirstRemoved = true;
	let isFirstAdded = true;

	for (const part of wordDiff) {
		if (part.removed) {
			let value = part.value;
			// Strip leading whitespace from the first removed part
			if (isFirstRemoved) {
				const leadingWs = value.match(/^(\s*)/)?.[1] || "";
				value = value.slice(leadingWs.length);
				removedLine += leadingWs;
				isFirstRemoved = false;
			}
			if (value) {
				removedLine += theme.inverse(value);
			}
		} else if (part.added) {
			let value = part.value;
			// Strip leading whitespace from the first added part
			if (isFirstAdded) {
				const leadingWs = value.match(/^(\s*)/)?.[1] || "";
				value = value.slice(leadingWs.length);
				addedLine += leadingWs;
				isFirstAdded = false;
			}
			if (value) {
				addedLine += theme.inverse(value);
			}
		} else {
			removedLine += part.value;
			addedLine += part.value;
		}
	}

	return { removedLine, addedLine };
}

export interface RenderDiffOptions {
	/** File path (unused, kept for API compatibility) */
	filePath?: string;
}

/**
 * Render a diff string with colored lines and intra-line change highlighting.
 * - Context lines: dim/gray
 * - Removed lines: red, with inverse on changed tokens
 * - Added lines: green, with inverse on changed tokens
 */
export function renderDiff(diffText: string, _options: RenderDiffOptions = {}): string {
	const lines = diffText.split("\n");
	const result: string[] = [];

	let i = 0;
	while (i < lines.length) {
		const line = lines[i];
		const parsed = parseDiffLine(line);

		if (!parsed) {
			result.push(theme.fg("toolDiffContext", line));
			i++;
			continue;
		}

		if (parsed.prefix === "-") {
			// Collect consecutive removed lines
			const removedLines: { lineNum: string; content: string }[] = [];
			while (i < lines.length) {
				const p = parseDiffLine(lines[i]);
				if (!p || p.prefix !== "-") break;
				removedLines.push({ lineNum: p.lineNum, content: p.content });
				i++;
			}

			// Collect consecutive added lines
			const addedLines: { lineNum: string; content: string }[] = [];
			while (i < lines.length) {
				const p = parseDiffLine(lines[i]);
				if (!p || p.prefix !== "+") break;
				addedLines.push({ lineNum: p.lineNum, content: p.content });
				i++;
			}

			// Only do intra-line diffing when there's exactly one removed and one added line
			// (indicating a single line modification). Otherwise, show lines as-is.
			if (removedLines.length === 1 && addedLines.length === 1) {
				const removed = removedLines[0];
				const added = addedLines[0];

				const { removedLine, addedLine } = renderIntraLineDiff(
					replaceTabs(removed.content),
					replaceTabs(added.content),
				);

				result.push(theme.fg("toolDiffRemoved", `-${removed.lineNum} ${removedLine}`));
				result.push(theme.fg("toolDiffAdded", `+${added.lineNum} ${addedLine}`));
			} else {
				// Show all removed lines first, then all added lines
				for (const removed of removedLines) {
					result.push(theme.fg("toolDiffRemoved", `-${removed.lineNum} ${replaceTabs(removed.content)}`));
				}
				for (const added of addedLines) {
					result.push(theme.fg("toolDiffAdded", `+${added.lineNum} ${replaceTabs(added.content)}`));
				}
			}
		} else if (parsed.prefix === "+") {
			// Standalone added line
			result.push(theme.fg("toolDiffAdded", `+${parsed.lineNum} ${replaceTabs(parsed.content)}`));
			i++;
		} else {
			// Context line
			result.push(theme.fg("toolDiffContext", ` ${parsed.lineNum} ${replaceTabs(parsed.content)}`));
			i++;
		}
	}

	return result.join("\n");
}

export type DiffRowKind = "context" | "removed" | "added";

export interface DiffRow {
	kind: DiffRowKind;
	lineNum: string;
	content: string;
}

export const BODY_JOINT = "└ ";
export const BODY_JOINT_WIDTH = 2;

export function parseDiffRows(diffText: string): { rows: DiffRow[]; added: number; removed: number } {
	const rows: DiffRow[] = [];
	let added = 0;
	let removed = 0;

	for (const line of diffText.split("\n")) {
		const parsed = parseDiffLine(line);
		if (!parsed) continue;

		const kind: DiffRowKind = parsed.prefix === "+" ? "added" : parsed.prefix === "-" ? "removed" : "context";
		if (kind === "added") added++;
		if (kind === "removed") removed++;
		rows.push({ kind, lineNum: parsed.lineNum.trim(), content: parsed.content });
	}

	return { rows, added, removed };
}

export class DiffRowsComponent implements Component {
	private readonly rows: DiffRow[];
	private readonly highlightedContents?: string[];

	constructor(rows: DiffRow[], filePath?: string) {
		this.rows = rows;
		const lang = filePath ? getLanguageFromPath(filePath) : undefined;
		if (lang) {
			this.highlightedContents = highlightCode(rows.map((row) => replaceTabs(row.content)).join("\n"), lang);
		}
	}

	invalidate(): void {}

	render(width: number): string[] {
		const numWidth = Math.max(1, ...this.rows.map((row) => row.lineNum.length));
		const indentWidth = Math.min(BODY_JOINT_WIDTH, Math.max(0, width));
		const regionWidth = Math.max(0, width - indentWidth);
		const indent = " ".repeat(indentWidth);

		return this.rows.map((row, index) => {
			if (regionWidth === 0) return indent;

			const marker = row.kind === "context" ? "  " : row.kind === "removed" ? " -" : " +";
			const color = row.kind === "context" ? "toolDiffContext" : "toolDiffText";
			const prefix = theme.fg(color, `${row.lineNum.padStart(numWidth, " ")}${marker}`);
			const content = this.highlightedContents?.[index] ?? theme.fg(color, replaceTabs(row.content));
			const styled = truncateToWidth(prefix + content, regionWidth, "");
			const padded = styled + " ".repeat(regionWidth - visibleWidth(styled));
			if (row.kind === "context") return indent + padded;

			return indent + theme.bg(row.kind === "removed" ? "toolDiffRemovedBg" : "toolDiffAddedBg", padded);
		});
	}
}

File: python/src/pi_mono/coding_agent/modes/interactive/components/diff.py
"""Colored diff rendering for interactive tool output."""

from __future__ import annotations

import difflib
import re
from typing import TypedDict

from pi_mono.coding_agent.modes.interactive.theme.theme import theme

_DIFF_LINE_RE = re.compile(r"^([-+\s])(\s*\d*)\s(.*)$")
_WORD_RE = re.compile(r"\S+|\s+")


class RenderDiffOptions(TypedDict, total=False):
    filePath: str


def _parse_diff_line(line: str) -> dict[str, str] | None:
    match = _DIFF_LINE_RE.match(line)
    if not match:
        return None
    return {"prefix": match.group(1), "lineNum": match.group(2), "content": match.group(3)}


def _replace_tabs(text: str) -> str:
    return text.replace("\t", "   ")


def _diff_words(old_content: str, new_content: str) -> list[tuple[str, str]]:
    old_tokens = _WORD_RE.findall(old_content)
    new_tokens = _WORD_RE.findall(new_content)
    matcher = difflib.SequenceMatcher(None, old_tokens, new_tokens)
    parts: list[tuple[str, str]] = []
    for tag, i1, i2, j1, j2 in matcher.get_opcodes():
        if tag == "equal":
            value = "".join(old_tokens[i1:i2])
            if value:
                parts.append(("equal", value))
        elif tag == "delete":
            value = "".join(old_tokens[i1:i2])
            if value:
                parts.append(("removed", value))
        elif tag == "insert":
            value = "".join(new_tokens[j1:j2])
            if value:
                parts.append(("added", value))
        else:
            removed = "".join(old_tokens[i1:i2])
            added = "".join(new_tokens[j1:j2])
            if removed:
                parts.append(("removed", removed))
            if added:
                parts.append(("added", added))
    return parts


def _render_intra_line_diff(old_content: str, new_content: str) -> tuple[str, str]:
    removed_line = ""
    added_line = ""
    is_first_removed = True
    is_first_added = True

    for part_tag, value in _diff_words(old_content, new_content):
        if part_tag == "removed":
            leading_ws_match = re.match(r"^(\s*)", value)
            leading_ws = leading_ws_match.group(1) if leading_ws_match else ""
            trimmed = value[len(leading_ws) :]
            if is_first_removed:
                removed_line += leading_ws
                is_first_removed = False
            if trimmed:
                removed_line += theme.inverse(trimmed)
        elif part_tag == "added":
            leading_ws_match = re.match(r"^(\s*)", value)
            leading_ws = leading_ws_match.group(1) if leading_ws_match else ""
            trimmed = value[len(leading_ws) :]
            if is_first_added:
                added_line += leading_ws
                is_first_added = False
            if trimmed:
                added_line += theme.inverse(trimmed)
        else:
            removed_line += value
            added_line += value

    return removed_line, added_line


def is_display_diff(diff_text: str) -> bool:
    for line in diff_text.split("\n"):
        if not line.strip():
            continue
        return _parse_diff_line(line) is not None
    return False


def render_diff(diff_text: str, _options: RenderDiffOptions | None = None) -> str:
    lines = diff_text.split("\n")
    result: list[str] = []
    index = 0

    while index < len(lines):
        line = lines[index]
        parsed = _parse_diff_line(line)

        if not parsed:
            result.append(theme.fg("toolDiffContext", line))
            index += 1
            continue

        if parsed["prefix"] == "-":
            removed_lines: list[dict[str, str]] = []
            while index < len(lines):
                current = _parse_diff_line(lines[index])
                if not current or current["prefix"] != "-":
                    break
                removed_lines.append(current)
                index += 1

            added_lines: list[dict[str, str]] = []
            while index < len(lines):
                current = _parse_diff_line(lines[index])
                if not current or current["prefix"] != "+":
                    break
                added_lines.append(current)
                index += 1

            if len(removed_lines) == 1 and len(added_lines) == 1:
                removed = removed_lines[0]
                added = added_lines[0]
                removed_line, added_line = _render_intra_line_diff(
                    _replace_tabs(removed["content"]),
                    _replace_tabs(added["content"]),
                )
                result.append(theme.fg("toolDiffRemoved", f"-{removed['lineNum']} {removed_line}"))
                result.append(theme.fg("toolDiffAdded", f"+{added['lineNum']} {added_line}"))
            else:
                for removed in removed_lines:
                    result.append(
                        theme.fg(
                            "toolDiffRemoved",
                            f"-{removed['lineNum']} {_replace_tabs(removed['content'])}",
                        )
                    )
                for added in added_lines:
                    result.append(
                        theme.fg(
                            "toolDiffAdded",
                            f"+{added['lineNum']} {_replace_tabs(added['content'])}",
                        )
                    )
        elif parsed["prefix"] == "+":
            result.append(
                theme.fg(
                    "toolDiffAdded",
                    f"+{parsed['lineNum']} {_replace_tabs(parsed['content'])}",
                )
            )
            index += 1
        else:
            result.append(
                theme.fg(
                    "toolDiffContext",
                    f" {parsed['lineNum']} {_replace_tabs(parsed['content'])}",
                )
            )
            index += 1

    return "\n".join(result)
